### PR TITLE
Fix issue #7

### PR DESCRIPTION
**Pull request by harikt**
Sorry @pmjones I wasn't able to get your commits and merge first. Your test case runs here. 

Running out of time.

### CODE REVIEW

**Comment by pmjones** [comment]
This one smells funny to me. The unsetting of the $key and then setting $params into itself. (I considered setting $params into itself earlier on, for what it's worth.)

Maybe the real solution is not to deal with this case at all, and instead suggest a shared $params ArrayObject that gets injected into both the dispatcher and the dispatchable object.

File: src/Dispatcher.php [hunk 1]
```diff
@@ -85,8 +85,16 @@ public function __construct(
...
     public function __invoke($params = [])
-    {
+    {        
         $object = $this->getObjectByParams($params);
+        // type cast params to array
+        $params = (array) $params;
+        if (! isset($params['params'])) {
+            $key = $this->getObjectParam();
+            // unset the controller, probably a closure
+            unset($params[$key]);
```
Comment (harikt) [inline src/Dispatcher.php:95]: @pmjones the unset of key is for in our case we try to cast via implode. But it can be removed.

Comment (pmjones) [inline src/Dispatcher.php:95]: Yeah, even so, we're setting $params into itself, and that just doesn't seem right somehow.

**Comment by harikt** [comment]
or may be we should do in `invokeMethod` , `invokeTrait` the remaining params. An array merge, not sure though.

And I don't get your way of injecting shared params ArrayObject :( .

File: src/Dispatcher.php (hunk 1 above)
Comment (harikt) [inline src/Dispatcher.php:95]: ok, what about in `invokeMethod` and `invokeTrait` setting the params we have in `$args`

**Comment by pmjones** [comment]
The idea would be you create a $params object (e.g. \Aura\Web\Request\Params) and pass it to the dispatcher, _and_ you construct Blog with it.  Then they both have the same object to work with; the dispatcher does not have to pass it to Blog at all.

**Comment by harikt** [comment]
For the `\Aura\Web\Request\Params` you need to set the params via `Aura.Router` . So does passing the object has any benefit ?

**Comment by pmjones** [comment]
Using a router, or even using \Aura\Web\Request\Params specifically, is not the point. The point is that the same object is passed to both the dispatcher and the blog object, so they both have access to the same things.

**Comment by harikt** [comment]
I will wait to see your implementation.

**Comment by harikt** [comment]
Probably the params mentioning at https://github.com/auraphp/Aura.Dispatcher#recursion-and-lazy-loading is not good then, for all the params are not setting back.

May be if $args is empty and $params not empty you set back. Just a thought .

**Comment by harikt** [comment]
``` php
if (empty($args) && ! empty($params)) {
    $args = $params;
}
```

in `invokeMethod` and `invokeClosure`

**Comment by harikt** [comment]
no it will not work, as I expected :) .

**Comment by pmjones** [comment]
It's interesting to see you trying all the things I tried. ;-)

**Comment by harikt** [comment]
he he he, I tried many stuffs yesterday, but didn't worked as I wished. And today morning seeing your way I thought the appending `params` to itself is a nice way. But you disagree with it :( .

**Comment by harikt** [comment]
anyway it is interesting to learn the inner working, so no issues at all :+1:  .

**Comment by pmjones** [comment]
I get you.

To be clear, the solution I committed does not append $params; it looks for a special `$_params` arg name, and sends `$params` to it directly.  I'm not really happy with that, since it makes `$_params` a magic arg name, and I want to avoid magic.  I may end up taking it out entirely.

However, something good came of this: we discovered that the dispatcher was passing a `null` when a param was missing, and that turns out to be a bad idea.

**Comment by harikt** [comment]
Probably we can document it like if you have arguments named pass it with the same name.

``` php
$params = [
    'controller' => 'blog',
    'params' => [
        'action' => 'read',
            'id' => 88,
    ]
];

$result = $dispatcher->__invoke($params);
```

**Comment by pmjones** [comment]
We are so dumb. If you as the user want to pass a param that matches an argument named 'params', set it into itself yourself.

``` php
<?php
$params = [
    'controller' => 'blog',
    'action' => 'read',
    'id' => 88,
]

$params['params'] =& $params;
?>
```

**Comment by harikt** [comment]
:)

**Pull request closed without merging**